WordPress/gutenberg
# Global styles: Allow indirect properties when unfiltered_html is not allowed

## What?


Fixes #45520

This PR adds a new `INDIRECT_PROPERTIES_METADATA` array to the theme JSON class, to store properties metadata for properties that are not directly output by `compute_style_properties`, but are instead used elsewhere in global styles output.

This array is then used in `remove_insecure_styles` to ensure that style values that will be used in other cases are validated against an indirect property. The main use case at the moment is for `blockGap` style values, which are output as either top margins or gap styles, depending on the layout type. For the purposes of validation, though, we can safely validate against a small list of gap properties.

## Why?


Over in #45520 it was flagged that the absence of `blockGap` in `PROPERTIES_METADATA` meant that the value was being stripped in situations where a user doesn't have the `unfiltered_html` capability and is saving global styles (e.g. for multisite regular admins). Hopefully, this PR should resolve that.

## How?


* Add an `INDIRECT_PROPERTIES_METADATA` array to the theme JSON class.
* Validate against this array in `remove_insecure_styles`.
* Add a Gutenberg version of `wp_filter_global_styles_post` (`gutenberg_filter_global_styles_post`) so that we can test this change in Gutenberg
* If the core versions of `wp_filter_global_styles_post` callbacks are registered, then de-register, and re-register the Gutenberg versions instead

## Testing Instructions





Note: a quick way to test this PR is to place the following in `lib/init.php` in the Gutenberg plugin, rather than worrying about changing user capabilities:

```
add_action( 'init', 'kses_init_filters' );
```

1. With this PR applied and with a user with `unfiltered_html` capability (e.g. a regular admin), you should be able to save root block spacing values in global styles, as well as block level spacing (e.g. set Buttons spacing, and axial row/column spacing for Social Icons).
2. Repeating the above with a user _without_ `unfiltered_html` capability (or enable the KSES filters with `add_action( 'init', 'kses_init_filters' );`) you should now still be able to save root block spacing values in global styles, and block level spacing.

Ensure tests pass:

```
npm run test:unit:php -- --filter WP_Theme_JSON_Gutenberg_Test
```

## Screenshots or screencast 

| Before (note value is reset when saved) | After (value is saved) |
| --- | --- |
| ![2022-12-08 11 32 33](https://user-images.githubusercontent.com/14988353/206327161-2e80181f-4a4f-4786-95a5-afe4603e7a2c.gif) | ![2022-12-08 11 28 44](https://user-images.githubusercontent.com/14988353/206326699-bd2f9cc5-1fd7-4bfa-b3eb-fd66f6c5c4d9.gif) |
